polars/polars-core/src/chunked_array/kernels/mod.rs

```diff
@@ -7,173 +7,56 @@ pub(crate) mod take_agg;
 #[cfg_attr(docsrs, doc(cfg(feature = "temporal")))]
 pub mod temporal;
 
-use polars_arrow::builder::BooleanArrayBuilder;
-
-use crate::chunked_array::builder::get_bitmap;
-use crate::datatypes::{
-    ArrowDataType, Float64Type, PolarsFloatType, PolarsNumericType, PolarsPrimitiveType,
-};
-use crate::prelude::AlignedVec;
-use arrow::array::{Array, ArrayData, ArrayRef, PrimitiveArray};
-use arrow::datatypes::{
-    Float32Type, Int16Type, Int32Type, Int64Type, Int8Type, UInt16Type, UInt32Type, UInt64Type,
-    UInt8Type,
-};
-use num::{Float, NumCast};
+use crate::datatypes::PolarsFloatType;
+use arrow::array::{Array, ArrayRef, BooleanArray, PrimitiveArray};
+use arrow::bitmap::Bitmap;
+use num::Float;
 use std::sync::Arc;
 
-pub(crate) unsafe fn transmute_array<S, T>(arr: &PrimitiveArray<S>) -> ArrayRef
-where
-    S: PolarsPrimitiveType,
-    T: PolarsPrimitiveType,
-{
-    let data = arr.data();
-    let buf = data.buffers().to_vec();
-    let child_data = data.child_data().to_vec();
-    let (null_count, null_bit_buffer) = get_bitmap(arr);
-    let new_data = ArrayData::new(
-        T::DATA_TYPE,
-        arr.len(),
-        Some(null_count),
-        null_bit_buffer,
-        data.offset(),
-        buf,
-        child_data,
-    );
-    Arc::new(PrimitiveArray::<T>::from(new_data))
-}
-
-pub(crate) unsafe fn transmute_array_from_dtype<S>(
-    arr: &PrimitiveArray<S>,
-    dtype: ArrowDataType,
-) -> ArrayRef
-where
-    S: PolarsPrimitiveType,
-{
-    use ArrowDataType::*;
-    match dtype {
-        UInt64 => transmute_array::<_, UInt64Type>(arr),
-        UInt32 => transmute_array::<_, UInt32Type>(arr),
-        UInt16 => transmute_array::<_, UInt16Type>(arr),
-        UInt8 => transmute_array::<_, UInt8Type>(arr),
-        Int64 => transmute_array::<_, Int64Type>(arr),
-        Int32 => transmute_array::<_, Int32Type>(arr),
-        Int16 => transmute_array::<_, Int16Type>(arr),
-        Int8 => transmute_array::<_, Int8Type>(arr),
-        _ => todo!(),
-    }
-}
-
-/// This function will panic if the conversion overflows. Don't use it to cast to a smaller size.
-pub(crate) fn cast_numeric<S, T>(arr: &PrimitiveArray<S>) -> ArrayRef
-where
-    S: PolarsNumericType,
-    T: PolarsNumericType,
-    T::Native: num::NumCast,
-    S::Native: num::NumCast,
-{
-    let vals = arr.values();
-    let (_null_count, null_bit_buffer) = get_bitmap(arr);
-    let av = vals
-        .iter()
-        .map(|v| num::cast::cast::<S::Native, T::Native>(*v).unwrap())
-        .collect::<AlignedVec<T::Native>>();
-    Arc::new(av.into_primitive_array::<T>(null_bit_buffer))
-}
-
 pub(crate) fn is_nan<T>(arr: &PrimitiveArray<T>) -> ArrayRef
 where
     T: PolarsFloatType,
     T::Native: Float,
 {
-    let vals = arr.values();
-    let (_, null_bit_buffer) = get_bitmap(arr);
-    let mut builder = BooleanArrayBuilder::new_no_nulls(vals.len());
+    let validity = arr.validity();
```

Comment:
arrow2's "nulls" is called a `validity`, as per arrow's own notation. A validity is a `Option<Bitmap>`. `Bitmap` has some methods to get the validity: `get_bit` and `.iter` are the most common ones.

In general, it is advantageous to branch the code as

```
if let Some(validity) = array.validity() {
   ...
} else {
...
}
```

since it allows faster branches that skip validities altogether.

Reply:
Clear, great that they implement `iter`. More idiomatic, and probably has no bound checking in safe code.

Reply:
the `is_nan` functions is also way more readable! :+1: